components/fxa-client/src/oauth.rs

```diff
@@ -206,12 +210,25 @@ impl FirefoxAccount {
         // In order to keep 1 and only 1 refresh token alive per client instance,
         // we also destroy the existing refresh token.
         if let Some(ref old_refresh_token) = self.state.refresh_token {
+            // Destroying a refresh token also destroys its associated device,
+            // grab the device information for replication later.
+            let device_info = self.get_current_device()?;
             if let Err(err) = self
                 .client
                 .destroy_oauth_token(&self.state.config, &old_refresh_token.token)
             {
                 log::warn!("Refresh token destruction failure: {:?}", err);
             }
+            if let Some(device_info) = device_info {
+                if let Err(err) = self.replace_device(
+                    &device_info.display_name,
+                    &device_info.device_type,
+                    &device_info.push_subscription,
+                    &device_info.available_commands,
+                ) {
+                    log::warn!("Device information restoration failed: {:?}", err);
+                }
```

Comment:
So if this does error out (or equivalently, if we crashed in between deleting the refresh token and re-creating the device record) is there some way for the app to detect that it no longer has a device record and needs to re-register push endpoints etc?

Reply:
I don't think it does, but I don't really expect this to be a problem. We can do regular re-registrations if we feel it becomes a problem later.